aspnetcore/grpc/migration.md

```diff
@@ -0,0 +1,87 @@
+---
+title: Migrating gRPC services from C Core to ASP.NET Core
+author: juntaoluo
+description: Learn how to move an existing C Core based gRPC app to run on top of ASP.NET Core stack.
+monikerRange: '>= aspnetcore-3.0'
+ms.author: johluo
+ms.date: 03/08/2019
+uid: grpc/migration
+---
+# Migrating gRPC services from C Core to ASP.NET Core
+
+By [John Luo](https://github.com/juntaoluo)
+
+Due to implementation of the underlying stack, not all features work in the same way between C Core based gRPC apps and ASP.NET Core based apps. This document highlights the key differences to keep in mind when migrating between the two stacks.
+
+## gRPC service implementation lifetime
+
+In the ASP.NET Core stack, gRPC services, by default, will be created with a Scoped lifetime as defined in the [Dependency Injection (DI)](xref:fundamentals/dependency-injection) document. In contrast, gRPC C Core by default binds to a service with a Singleton lifetime.
+
+A Scoped lifetime allows the service implementation to resolve other services with Scoped lifetimes, such as DBContext, from the DI container through constructor injection. Since a new instance of the service implementation is constructed for each request, it is no longer possible to share state between requests via instance members on the implementation type. Instead, the expectation is to store shared states in a Singleton service in the DI container and resolve it in the constructor of the gRPC service implementation.
+
+### Add a singleton service
+
+To facilitate the transition from gRPC C Core implementation to ASP.NET Core, it is possible to change the service lifetime of the service implementation from Scoped to Singleton. This involves adding an instance of the service implementation to the DI container:
```

Comment:
Should mention that there are two options: moving state to a third-party interface and injecting that vs making the service singleton.

Reply:
That was in the previous section:
> Instead, the expectation is to store shared states in a Singleton service in the DI container and resolve it in the constructor of the gRPC service implementation.